src/main/java/com/google/sps/servlets/UploadServlet.java

```diff
@@ -0,0 +1,150 @@
+    long timestamp = System.currentTimeMillis();
+
+    // Loop to see if that user already exists
+    boolean datasetPresent = false;
+    boolean userPresent = false;
+    boolean newDataset = true;  // Assume this is a new dataset being created
+    for (Entity entity : results.asIterable()) {
+      String email = (String) entity.getProperty("user-email");
+      String dataset = (String) entity.getProperty("dataset-name");
+
+      if (email.equals(userEmail)) {
+        userPresent = true;
+        if (dataset.equals(datasetName)) {
+          // Do nothing, use this existing dataset
+          datasetPresent = true;
+          newDataset = false;
+        }
+      }
+    }
+    
+    // If this user already has that dataset present
+    if (userPresent && datasetPresent) {
+      response.sendRedirect("/invalid.html");
+    }
+
+
+    if (!datasetPresent) {
+      /* Upload images from form to GCP bucket */
+
+      int imageCount = 0;
+
+      // The ID of your GCP project
+      String projectId = "step-2020-johndallard";
+
+      // The ID of your GCS bucket
+      String bucketName = "embedding-visualizer-bucket";
+
+      Storage storage = StorageOptions.newBuilder().setProjectId(projectId).build().getService();
+
+      // Create new directory within user's subdirectory for this dataset
+      String userDirectory = userEmail + "/";
+      String newDatasetDir = userDirectory + datasetName + "/";
+      String userImagesDir = newDatasetDir + "original_images/";
+      String userThumbnailsDir = newDatasetDir + "thumbnails/";
+      String userSpritesheetDir = newDatasetDir + "spritesheets/";
+      String userEmbeddingsDir = newDatasetDir + "embeddings/";
+      String[] directories = {userImagesDir, userThumbnailsDir, userSpritesheetDir, userEmbeddingsDir};
+
+      BlobId blobId = BlobId.of(bucketName, userDirectory);
+      BlobInfo blobInfo = BlobInfo.newBuilder(blobId).build();
+      storage.create(blobInfo);
+
+      // Create new subdirectories for the new dataset
+      for (int i = 0; i < directories.length; i++) {
+        blobId = BlobId.of(bucketName, directories[i]);
+        blobInfo = BlobInfo.newBuilder(blobId).build();
+        storage.create(blobInfo);
+      }
+
+      Collection<Part> parts = request.getParts();
+      
+      for (Part part : parts) {
+        if (!"file-upload-dialog".equals(part.getName())) continue;
+        InputStream inputStream = part.getInputStream();
+
+        String fileName = Paths.get(part.getSubmittedFileName()).getFileName().toString(); // MSIE fix.
+        //The ID of your GCS object
+        String objectName = userImagesDir + fileName;
+
+        blobId = BlobId.of(bucketName, objectName);
+        blobInfo = BlobInfo.newBuilder(blobId).build();
+        storage.create(blobInfo, inputStream);
+        imageCount++;
+      }
+    
+      /* Store metadata in Datastore */
+      Entity datasetEntity = new Entity("MetaData");
```

Comment:
I think it will be neater to create a separate class for datastore operations (let me talk to you offline tomorrow morning about this), but for this PR, let's leave it as it is.

Reply:
As we discussed, I'll save this for a separate PR after this, including my own tests